### ISSUE TITLE
Raw messages in Whisper

### ISSUE THREAD

**Issue by sprusr** [ethereum/go-ethereum#16708]
Moving this to an issue after after a discussion with @gballet on Gitter.

## Proposal

At present, when sending a message via Whisper, all crypto is done by the node. While this does provide us with a level of security when using a local node, many users of dApps use tools such as MetaMask which connect to a public node to interact with Ethereum.

Users don't particularly want to be handing over their keys to a public node (even if generating for a single-use), in the same way that this isn't the case for sending transactions - instead the transaction is signed locally before being submitted to the node with `eth_sendRawTransaction`.

In a similar fashion it would be useful to have something like a `shh_sendRawMessage` method.

## Problems

As @frozeman pointed out when I first discussed this with him, Whisper also requires that the user listen for messages intended for them. Receiving raw messages would probably not be as trivial as sending them, although it is certainly possible.

@gballet also raised concerns about the leaking of metadata when using a public node to forward Whisper messages onto the network.

## Other Thoughts

While there are security concerns where the sender wishes to remain entirely anonymous, the majority of uses cases do not require this level of absolute privacy. So long as the payload remains secret, this is enough for most users.

The addition of such a feature could pave the way for MetaMask, Infura, etc. adding Whisper support (see https://github.com/INFURA/infura/issues/19). That would make it much more feasible to use Whisper in dApps.

**Comment by gballet** [comment]
Thanks for moving the discussion right here. Right now, the best approach is to run a local `wnode` and connect to the network (possibly with a very limited filter, in order to reduce traffic) instead of using infura directly. This being said, there is indeed a case for sending raw transactions. I'm perfectly happy with having an initial "message in a bottle" proof of concept, where it is only possible to send messages this way.

Polling the server is a problem, given the amount of traffic and the short time to live of messages in memory. It would be possible to poll for a specific topic, but this implies a trade-off with respect to metadata security.

Is there a use case when the user would want to send messages and not read them?

**Comment by fukumitsu76** [comment]
2018年5月9日(水) 19:34 Guillaume Ballet <notifications@github.com>:

**Comment by fukumitsu76** [comment]
2018年5月9日(水) 19:34 Akiyoshi Fukumitsu <fukumitsu@hivelocity.co.jp>:

**Comment by frozeman** [comment]
The pub sub would be a great fit here, as it reduces unnecessary polling.
At the same time I agree, giving the node the topic is a security concern.

@gballet to decide a message encrypted for a specific address, does the node just tries to decide it with all its stored private keys, or is in those messages the “to” always present?

**Comment by gballet** [comment]
> The pub sub would be a great fit here, as it reduces unnecessary polling.

@frozeman I don't know the details of the pub-sub implementation; it seems to me however that you would need to keep a list of sockets open and I see a scalability issue here: not only must the server send whisper messages over devp2p, but it must maintain a list of its clients and send the same data as many times over `ws://` as there are clients. As you point out, you could indeed reduce the traffic by only registering for some specific topics, but that would leak metadata.

On the plus side, the overwhelming traffic would be an improvement in plausible deniability compared to web3; the bad part is that even more code needs to be duplicated between the Go client and the Javascript client.

This is becoming messy and I'm not necessary this is worth the work. Hence my question: is there a use case where sending messages and not reading them with the same method would make sense ?

**Comment by sprusr** [comment]
Yeah, I don't think there are many use cases where one would send a message without expecting a response. Therefore, it does not make sense to add only sending in this way.

I understand your concerns about achieving darkness, however I think for many users this is not a priority, as long as the contents of the message is secret. Notifying a public node of which topics a user wishes to receive will leak a certain amount of metadata, but for users where this is a concern they may simply run their own local node and not use public ones.

Perhaps it would help to explain the context I'm considering this all in: from the point of view of a browser-based dApp, to do things such as off-chain negotiation before a contract-based agreement. As a dApp developer, I want to be able to use Whisper in my application without the user having to install a separate browser (e.g. with MetaMask).
